pkg/occlient/occlient.go

```diff
@@ -1518,46 +1424,174 @@ func (c *Client) GetOnePodFromSelector(selector string) (*corev1.Pod, error) {
 	return &pods.Items[0], nil
 }
 
-// RsyncPath copies local directory to directory in running Pod.
-func (c *Client) RsyncPath(localPath string, targetPodName string, targetPath string) (string, error) {
-	log.Debugf("Syncing %s to pod %s:%s", localPath, targetPodName, targetPath)
-
-	// TODO: do this without using 'oc' binary
-	args := []string{
-		"rsync",
-		localPath,
-		fmt.Sprintf("%s:%s", targetPodName, targetPath),
-		"--exclude", ".git",
-		"--no-perms",
+// CopyFile copies localFile to the directory in running Pod.
+// copyFiles is list of changed files captured during `odo watch`
+// localFile is ignored if copyFiles is set
+func (c *Client) CopyFile(localPath string, targetPodName string, targetPath string, copyFiles []string) error {
+	if len(copyFiles) != 0 || string(localPath[len(localPath)-1]) == "/" {
+		log.Debugf("Copying directory %s to pod %s:%s", copyFiles, targetPodName, targetPath)
+	} else {
+		log.Debugf("Copying file %s to pod %s:%s", localPath, targetPodName, targetPath)
 	}
 
-	output, err := c.runOcComamnd(&OcCommand{args: args})
+	dest := targetPath + "/" + path.Base(localPath)
+	reader, writer := io.Pipe()
+
+	go func() {
+		defer writer.Close()
+		err := makeTar(localPath, dest, writer, copyFiles)
+		if err != nil {
+			os.Exit(-1)
+		}
+
+	}()
+
+	var cmdArr []string
+
+	if len(copyFiles) != 0 || string(localPath[len(localPath)-1]) == "/" {
+		cmdArr = []string{"tar", "xf", "-", "-C", targetPath, "--strip", "1"}
+	} else {
+		cmdArr = []string{"tar", "xf", "-", "-C", targetPath}
+	}
+	err := c.ExecCMDInContainer(targetPodName, cmdArr, writer, writer, reader, false)
 	if err != nil {
-		return "", err
+		return err
 	}
+	return nil
```

Comment:
I don't see a sync check, for this function, should we wait until the go routine is finished before proceeding?

Reply:
I am confused little bit, here;s the reference: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/cp.go#L235